lib/timers.js

```diff
@@ -1,52 +1,150 @@
+// queue), and the object map lookup of a specific list by the duration of
+// timers within (or creation of a new list).
+// However, these operations combined have shown to be trivial in comparison to
+// other alternative timers architectures.
+
+
+// Object maps containing linked lists of timers, keyed and sorted by their
+// duration in milliseconds.
+//
+// The difference between these two objects is that the former contains timers
+// that will keep the process open if they are the only thing left, while the
+// latter will not.
 //
-// Because often many sockets will have the same idle timeout we will not
-// use one timeout watcher per item. It is too much overhead.  Instead
-// we'll use a single watcher for all sockets with the same timeout value
-// and a linked list. This technique is described in the libev manual:
-// http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#Be_smart_about_timeouts
-
-// Object containing all lists, timers
-// key = time in milliseconds
-// value = list
-var lists = {};
-
-
-// call this whenever the item is active (not idle)
-// it will reset its timeout.
-// the main function - creates lists on demand and the watchers associated
-// with them.
+// - key = time in milliseconds
+// - value = linked list
+const refedLists = {};
+const unrefedLists = {};
+
+
+// Schedule or re-schedule a timer.
+// The item must have been enroll()'d first.
 exports.active = function(item) {
-  const msecs = item._idleTimeout;
-  if (msecs < 0 || msecs === undefined) return;
+  insert(item, false);
+};
 
-  item._idleStart = Timer.now();
+// Internal APIs that need timeouts should use `_unrefActive()` instead of
+// `active()` so that they do not unnecessarily keep the process open.
+exports._unrefActive = function(item) {
+  insert(item, true);
+};
 
-  var list;
 
-  if (lists[msecs]) {
-    list = lists[msecs];
-  } else {
-    list = new Timer();
-    list.start(msecs, 0);
+// The underlying logic for scheduling or re-scheduling a timer.
+//
+// Appends a timer onto the end of an existing timers list, or creates a new
+// TimerWrap backed list if one does not already exist for the specified timeout
+// duration.
+function insert(item, unrefed) {
+  const msecs = item._idleTimeout;
+  if (msecs < 0 || msecs === undefined) return;
 
+  item._idleStart = TimerWrap.now();
+
+  const lists = unrefed === true ? unrefedLists : refedLists;
+
+  // Use an existing list if there is one, otherwise we need to make a new one.
+  var list = lists[msecs];
+  if (!list) {
+    debug('no %d list was found in insert, creating a new one', msecs);
+    // Make a new linked list of timers, and create a TimerWrap to schedule
+    // processing for the list.
+    list = {
+      _idleNext: null, // Create the list object with the linkedlist properties
+      _idlePrev: null, // to prevent any unnecessary hidden class changes.
+      _timer: new TimerWrap(),
+      _unrefed: unrefed,
+      msecs: msecs
+    };
```

Comment:
It will add just a hair of performance (hence why I added `TickObject` in `src/node.js`).